File: cs/src/Ice/Exception.cs
// **********************************************************************
//
// Copyright (c) 2003-2009 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using System.Diagnostics;

namespace IceInternal
{
    public class Ex
    {
        public static void throwUOE(string expectedType, string actualType)
        {
            throw new Ice.UnexpectedObjectException(
                        "expected element of type `" + expectedType + "' but received '" + actualType,
                        actualType, expectedType);
        }
    }
}

namespace Ice
{

    public abstract class Exception : System.Exception, System.ICloneable
    {
        public object Clone()
        {
            return MemberwiseClone();
        }

        public Exception() {}
        public Exception(System.Exception ex) : base("", ex) {}
        public abstract string ice_name();
        public override string ToString()
        {
            //
            // This prints the exception Java style. That is, the outermost
            // exception, "Caused by:" to the innermost exception. The
            // stack trace is not nicely indented as with Java, but
            // without string parsing (perhaps tokenize on "\n"), it
            // doesn't appear to be possible to reformat it.
            //
            System.IO.StringWriter sw = new System.IO.StringWriter();
            IceUtilInternal.OutputBase op = new IceUtilInternal.OutputBase(sw);
            op.setUseTab(false);
            op.print(GetType().FullName);
            op.inc();
            IceInternal.ValueWriter.write(this, op);
            sw.Write("\n");
            sw.Write(StackTrace);

            System.Exception curr = InnerException;
            while(curr != null)
            {
                sw.Write("\nCaused by: ");
                sw.Write(curr.GetType().FullName);
                if(!(curr is Ice.Exception))
                {
                    sw.Write(": ");
                    sw.Write(curr.Message);
                }
                sw.Write("\n");
                sw.Write(curr.StackTrace);
                curr = curr.InnerException;
            }

            return sw.ToString();
        }
    }

    public abstract class LocalException : Exception
    {
        public LocalException() {}
        public LocalException(System.Exception ex) : base(ex) {}
    }

    public abstract class SystemException : Exception
    {
        public SystemException() {}
        public SystemException(System.Exception ex) : base(ex) {}
    }

    public abstract class UserException : Exception
    {
        public UserException() {}
        public UserException(System.Exception ex) : base(ex) {}
        public abstract void write__(IceInternal.BasicStream os__);
        public abstract void read__(IceInternal.BasicStream is__, bool rid__);

        public virtual void write__(Ice.OutputStream outS__)
        {
            Debug.Assert(false);
        }

        public virtual void read__(Ice.InputStream inS__, bool rid__)
        {
            Debug.Assert(false);
        }

        public virtual bool usesClasses__()
        {
            return false;
        }
    }

}

namespace IceInternal
{

    public class LocalExceptionWrapper : Ice.Exception
    {
        public LocalExceptionWrapper(Ice.LocalException ex, bool retry)
        {
            _ex = ex;
            _retry = retry;
        }

        public LocalExceptionWrapper(LocalExceptionWrapper ex)
        {
            _ex = ex.get();
            _retry = ex._retry;
        }

        public override string ice_name()
        {
            return _ex.ice_name();
        }

        public Ice.LocalException get()
        {
            return _ex;
        }

        //
        // If true, always repeat the request. Don't take retry settings
        // or "at-most-once" guarantees into account.
        //
        // If false, only repeat the request if the retry settings allow
        // to do so, and if "at-most-once" does not need to be guaranteed.
        //
        public bool retry()
        {
            return _retry;
        }

        public static void throwWrapper(System.Exception ex)
        {
            if(ex is Ice.UserException)
            {
                throw new LocalExceptionWrapper(new Ice.UnknownUserException(((Ice.UserException)ex).ice_name()), false);
            }

            if(ex is Ice.LocalException)
            {
                if(ex is Ice.UnknownException ||
                   ex is Ice.ObjectNotExistException ||
                   ex is Ice.OperationNotExistException ||
                   ex is Ice.FacetNotExistException)
                {
                    throw new LocalExceptionWrapper((Ice.LocalException)ex, false);
                }
                throw new LocalExceptionWrapper(new Ice.UnknownLocalException(((Ice.LocalException)ex).ice_name()), false);
            }
            throw new LocalExceptionWrapper(new Ice.UnknownException(ex.GetType().FullName), false);
        }
        
        private Ice.LocalException _ex;
        private bool _retry;
    }

}

File: js/src/Ice/Exception.js
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

(function(){
    var global = this;
    
    require("Ice/Class");
    
    var Slice = global.Slice || {};
    var Ice = global.Ice || {};

    var defineClass = Ice.__defineClass;
    
    //
    // Ice.Exception
    //
    var Exception = defineClass(Error, {
        __init__: function(cause)
        {
            this.ice_cause = cause;
        },
        ice_name: function()
        {
            return "Ice::Exception";
        },
        toString: function()
        {
            return this.ice_name();
        }
    });
    
    Exception.captureStackTrace = function(object)
    {
        var stack = new Error().stack;
        
        var formattedStack;
        
        //
        // In IE 10 and greater the stack will be filled once the Error is throw
        // we don't need to do anything.
        //
        if(stack !== undefined)
        {

            var name =  object.ice_name ? object.ice_name().replace("::", ".") : "";
            Object.defineProperty(object, "stack", {
                get: function(){
                    if(formattedStack === undefined)
                    {
                        //
                        // Format the stack
                        //
                        var lines = stack.split("\n");
                        for(var i = 0; i < lines.length; ++i)
                        {
                            if(lines[i].indexOf(name) !== -1)
                            {
                                if(i < lines.length)
                                {
                                    lines = lines.slice(i + 1);
                                    break;
                                }
                            }
                        }
                        formattedStack = name + ": " + object.message + "\n"; 
                        formattedStack += lines.join("\n");
                    }
                    return formattedStack;
                }
            });
        }
    };

    Ice.Exception = Exception;

    //
    // Ice.LocalException
    //
    var LocalException = defineClass(Exception, {
        __init__: function(cause)
        {
            Exception.call(this, cause);
            Exception.captureStackTrace(this);
        },
        ice_name: function()
        {
            return "Ice::LocalException";
        }
    });
    
    Ice.LocalException = LocalException;
    
    Slice.defineLocalException = function(constructor, base, name)
    {
        var ex = constructor;
        ex.prototype = new base();
        ex.prototype.constructor = ex;
        ex.prototype.ice_name = function()
        {
            return name;
        };
        return ex;
    };

    //
    // Ice.UserException
    //
    var UserException = defineClass(Exception, {
        __init__: function(cause)
        {
            Exception.call(this, cause);
            Exception.captureStackTrace(this);
        },
        ice_name: function()
        {
            return "Ice::UserException";
        },
        __write: function(os)
        {
            os.startWriteException(null);
            __writeImpl(this, os, this.__mostDerivedType());
            os.endWriteException();
        },
        __read: function(is)
        {
            is.startReadException();
            __readImpl(this, is, this.__mostDerivedType());
            is.endReadException(false);
        }
    });
    Ice.UserException = UserException;
    
    //
    // Private methods
    //

    var __writeImpl = function(obj, os, type)
    {
        //
        // The __writeImpl method is a recursive method that goes down the
        // class hierarchy to marshal each slice of the class using the 
        // generated __writeMemberImpl method.
        //

        if(type === undefined || type === UserException)
        {
            return; // Don't marshal anything for Ice.UserException
        }

        os.startWriteSlice(type.__id, -1, type.__parent === UserException);
        if(type.prototype.__writeMemberImpl)
        {
            type.prototype.__writeMemberImpl.call(obj, os);
        }
        os.endWriteSlice();
        __writeImpl(obj, os, type.__parent);
    };

    var __readImpl = function(obj, is, type)
    {
        //
        // The __readImpl method is a recursive method that goes down the
        // class hierarchy to marshal each slice of the class using the 
        // generated __readMemberImpl method.
        //

        if(type === undefined || type === UserException)
        {
            return; // Don't marshal anything for UserException
        }

        is.startReadSlice();
        if(type.prototype.__readMemberImpl)
        {
            type.prototype.__readMemberImpl.call(obj, is);
        }
        is.endReadSlice();
        __readImpl(obj, is, type.__parent);
    };
    
    var __writePreserved = function(os)
    {
        //
        // For Slice exceptions which are marked "preserved", the implementation of this method
        // replaces the Ice.Object.prototype.__write method.
        //
        os.startWriteException(this.__slicedData);
        __writeImpl(this, os, this.__mostDerivedType());
        os.endWriteException();
    };

    var __readPreserved = function(is)
    {
        //
        // For Slice exceptions which are marked "preserved", the implementation of this method
        // replaces the Ice.Object.prototype.__read method.
        //
        is.startReadException();
        __readImpl(this, is, this.__mostDerivedType());
        this.__slicedData = is.endReadException(true);
    };
    
    Slice.defineUserException = function(constructor, base, name, writeImpl, readImpl, preserved)
    {
        var ex = constructor;
        ex.__parent = base;
        ex.prototype = new base();
        ex.__id = "::" + name;
        ex.prototype.ice_name = function()
        {
            return name;
        };
        
        ex.prototype.constructor = ex;
        ex.prototype.__mostDerivedType = function() { return ex; };
        if(preserved)
        {
            ex.prototype.__write = __writePreserved;
            ex.prototype.__read = __readPreserved;
        }
        ex.prototype.__writeMemberImpl = writeImpl;
        ex.prototype.__readMemberImpl = readImpl;
        return ex;
    };

    global.Slice = Slice;
    global.Ice = Ice;
}());
